### PR TITLE
Handle panic during validating admission webhook admission 

### PR DESCRIPTION
**Pull request by deads2k**
Validating admission webhook evaluation can fail, if uncaught this
crashes a kube-apiserver.  Add handling to catch panic while preserving
the behavior of "must not fail".

/kind bug
/priority important-soon
/sig api-machinery

```release-note
Panics while calling validating admission webhook are caught and honor the fail open or fail closed setting.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 1]
```diff
@@ -259,6 +273,8 @@ func (d *validatingDispatcher) callHook(ctx context.Context, h *v1.ValidatingWeb
 			klog.Warningf("Failed to set admission audit annotation %s to %s for validating webhook %s: %v", key, v, h.Name, err)
 		}
 	}
+	// this does bad things in general
+	panic("foo")
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:277]: this would require some mocking of `callHook` to wrap it, but since it would only be for a single instance, it's probably tolerable.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/plugin_test.go [hunk 2]
```diff
@@ -281,3 +283,72 @@ func TestValidateWebhookDuration(ts *testing.T) {
 		})
 	}
 }
+
+// TestValidatePanicHandling tests that panics should not escape the dispatcher
+func TestValidatePanicHandling(t *testing.T) {
+	// this seems unlikely to do good things during unit tests generally
+	utilruntime.ReallyCrash = false
```
[the remark this answers is not in the document]
Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/plugin_test.go:290]: this test is using this trick

https://github.com/kubernetes/kubernetes/pull/108455/files#diff-628f44e1ce35432138981f75ef30ef38570398e3772823f517215c8bc85cb8a0R312-R317

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 3]
```diff
@@ -100,11 +100,25 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
...
 	wg := sync.WaitGroup{}
-	errCh := make(chan error, len(relevantHooks))
+	errCh := make(chan error, 2*len(relevantHooks)) // double the length to handle extra errors for panics in the gofunc
 	wg.Add(len(relevantHooks))
 	for i := range relevantHooks {
 		go func(invocation *generic.WebhookInvocation, idx int) {
-			defer wg.Done()
+			ignoreClientCallFailures := false
+			defer func() {
+				defer wg.Done()
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:109]: defer inside defer is really hard to follow... is there a reason not to keep the `defer wg.Done()` where it was?

Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:109]: > defer inside defer is really hard to follow... is there a reason not to keep the defer wg.Done() where it was?

Calling Done of the wait group causes the parent thread to progress and close the error channel that this defer is attempting to write to.  This causes a race between the defer go func and the parent go func over which happens first.

If memory serves, it is possible to order the defer statement and gain a guaranteed happens-before, but that seemed more likely to be prone to future refactoring failures. I'm willing to structure it either way or add a comment since the reasoning is not obvious on a top to bottom read.

Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:109]: When you get it wrong `go test k8s.io/apiserver/pkg/admission/plugin/webhook/validating -run "TestValidatePanicHandling" --race` produces

> --- FAIL: TestValidatePanicHandling (0.32s)
    testing.go:1152: race detected during execution of test

But I haven't figured out how to merge the tests and depending on, "be careful to add these defers in the right order" wasn't clearly better to me than "defer inside of defer ensures correct ordering".  I'm flexible about factorization if your preference is to rely on the order in which defers are added, which is apparently called in reverse order (not the forward order that I would naively expect).

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:109]: non-nested defers called in reverse order reads more easily to me

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 4]
```diff
@@ -100,20 +100,53 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
...
 	for i := range relevantHooks {
 		go func(invocation *generic.WebhookInvocation, idx int) {
+			ignoreClientCallFailures := false
+			hookName := "unknown"
+			versionedAttr := versionedAttrs[invocation.Kind]
+			// The ordering of these two defers is critical.  The wg.Done will release the parent go func to close the errCh
+			// that is used by the second defer to report errors.  The recovery and error reporting must be done first.
 			defer wg.Done()
+			defer func() {
+				if r := recover(); r != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:114]:
```suggestion
				recover()
```

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 5]
```diff
@@ -100,20 +100,52 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
...
+				recover()
+			}()
+			defer utilruntime.HandleCrash(
+				func(r interface{}) {
+					if r == nil {
+						return
+					}
+					if ignoreClientCallFailures {
+						// if failures are supposed to ignored, ignore it
+						klog.Warningf("Panic calling webhook, failing open %v: %v", invocation.Webhook.GetName(), r)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:126]: isn't invocation.Webhook.GetName() the same as `hookName`? did you mean GetConfigurationName()?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 6]
```diff
@@ -100,20 +100,52 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
...
+						admissionmetrics.Metrics.ObserveWebhookFailOpen(ctx, hookName, "validating")
+						key := fmt.Sprintf("%sround_0_index_%d", ValidatingAuditAnnotationFailedOpenKeyPrefix, idx)
+						value := hookName
+						if err := versionedAttr.Attributes.AddAnnotation(key, value); err != nil {
+							klog.Warningf("Failed to set admission audit annotation %s to %s for validating webhook %s: %v", key, value, hookName, err)
+						}
+						return
+					}
+					// this ensures that the admission request fails and a message is provided.
+					errCh <- apierrors.NewInternalError(fmt.Errorf("ValidatingAdmissionWebhook/%v has panicked: %v", invocation.Webhook.GetName(), r))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:136]: isn't invocation.Webhook.GetName() the same as `hookName`? did you mean GetConfigurationName()?

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go [hunk 7]
```diff
@@ -100,20 +100,52 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
...
 	for i := range relevantHooks {
 		go func(invocation *generic.WebhookInvocation, idx int) {
+			ignoreClientCallFailures := false
+			hookName := "unknown"
+			versionedAttr := versionedAttrs[invocation.Kind]
+			// The ordering of these two defers is critical.  The wg.Done will release the parent go func to close the errCh
+			// that is used by the second defer to report errors.  The recovery and error reporting must be done first.
 			defer wg.Done()
+			defer func() {
+				// HandleCrash has already called the crash handlers and it  has  been configured to utilruntime.ReallyCrash
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:114]: > nit: weird double spaces in comments here and above

I miss my double spaces after periods

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go (hunk 6 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go:136]: > isn't invocation.Webhook.GetName() the same as hookName? did you mean GetConfigurationName()?

other errors here use hookName. Updated for variable instead of method.

File: staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go [hunk 8]
```diff
@@ -691,6 +691,97 @@ func NewNonMutatingTestCases(url *url.URL) []ValidatingTest {
...
+				ClientConfig:            ccfgSVC("internalErr"),
+				Rules:                   matchEverythingRules,
+				NamespaceSelector:       &metav1.LabelSelector{},
+				ObjectSelector:          &metav1.LabelSelector{},
+				FailurePolicy:           &policyFail,
+				AdmissionReviewVersions: []string{"v1beta1"},
+			}},
+			ExpectStatusCode: http.StatusInternalServerError,
+			ExpectAllow:      false,
+		},
```
[the remark this answers is not in the document]
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testing/testcase.go:782]: > no panic-related error message here?

added

**Comment by liggitt** [comment]
mm... looks like the unit test needs to tolerate whichever message comes first:

plugin_test.go:332: match & fail (but fail because fail closed): expected an error saying "ValidatingAdmissionWebhook/internalErr B has panicked: Start panicking!", but got Internal error occurred: ValidatingAdmissionWebhook/internalErr C has panicked: Start panicking!
